pkg/networkservice/metrics/server.go

```diff
@@ -0,0 +1,182 @@
+	metricsRecieved int
+	collectorCancel context.CancelFunc
+	Server
+}
+
+type connectionInfo struct {
+	connectionID string
+	index        uint32
+	metrics      map[string]string
+	ifName       string
+}
+
+func (s *metricsServer) GetMetricsReceived() (result int) {
+	<-s.executor.AsyncExec(func() {
+		result = s.metricsRecieved
+	})
+	return
+}
+
+func (s *metricsServer) Request(ctx context.Context, request *networkservice.NetworkServiceRequest) (*networkservice.Connection, error) {
+	conf := vppagent.Config(ctx)
+	if conf == nil {
+		return nil, errors.New("VPPAgent config is missing")
+	}
+	ifaces := conf.GetVppConfig().GetInterfaces()
+	if len(ifaces) == 0 {
+		return nil, errors.New("VPPAgent config should contain at least one interface")
+	}
+
+	index := request.GetConnection().GetPath().GetIndex()
+
+	conn, err := next.Server(ctx).Request(ctx, request)
+	if err == nil {
+		<-s.executor.AsyncExec(func() {
+			if len(s.connections) == 0 {
+				// Start collector go routine.
+				ctx, cancel := context.WithCancel(context.Background())
+				s.collectorCancel = cancel
+				go s.collect(ctx)
+			}
+			info := s.connections[conn.GetId()]
+			if info == nil {
+				info = &connectionInfo{
+					connectionID: conn.GetId(),
+					index:        conn.GetPath().GetIndex(),
+					metrics:      map[string]string{},
+					ifName:       ifaces[0].Name,
+				}
+				s.connections[conn.GetId()] = info
+				// Store interface name to store values into
+				s.vppConnections[info.ifName] = info.connectionID
+			}
+			// Update connection with metrics
+			if len(info.metrics) > 0 && len(conn.GetPath().GetPathSegments()) > int(index) {
+				conn.GetPath().GetPathSegments()[index].Metrics = info.metrics
+			}
+		})
+	}
+	return conn, err
+}
+
+func (s *metricsServer) Close(ctx context.Context, conn *networkservice.Connection) (*empty.Empty, error) {
+	s.executor.AsyncExec(func() {
+		info := s.connections[conn.GetId()]
+		if info != nil {
+			delete(s.connections, info.connectionID)
+			delete(s.vppConnections, info.ifName)
+		}
+		if len(s.connections) == 0 && s.collectorCancel != nil {
+			s.collectorCancel()
+			s.collectorCancel = nil
+		}
+	})
+	return next.Server(ctx).Close(ctx, conn)
+}
+
+// NewServer creates a new metrics collector instance
+func NewServer(interval time.Duration, vppClient configurator.StatsPollerServiceClient) networkservice.NetworkServiceServer {
+	rv := &metricsServer{
+		connections:    map[string]*connectionInfo{},
+		vppConnections: map[string]string{},
+		interval:       interval,
+		vppClient:      vppClient,
+	}
+	return rv
+}
+
+func (s *metricsServer) collect(ctx context.Context) {
+	req := &configurator.PollStatsRequest{
+		PeriodSec: uint32(s.interval.Seconds()),
+	}
+	stream, err := s.vppClient.PollStats(ctx, req)
+	if err != nil {
+		logrus.Errorf("MetricsCollector: PollStats err: %v", err)
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(s.interval):
```

Comment:
I believe this case will never be used, because `time.After` creates and starts new timer with given interval and `select` will pick the `default` case for any given interval (maybe even for 0).

Reply:
Fixed to have two cases to select cancel timeout and interval. Will remove this stuff after 0-intervals will be available. Thanks.

Reply:
Updated to use new vpp-agent functionality.